### ISSUE TITLE
network_cli perfomance is slow

### ISSUE THREAD

**Issue by vaneuk** [ansible/ansible#42058]
##### SUMMARY

When network modules are used with network_cli connection plugin the perfomance is slow. Task execution time is growing almost linearly with number of hosts.

##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME

network_cli

##### ANSIBLE VERSION

```
ansible 2.5.2
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/srv/ansible_modules']
  ansible python module location = /usr/local/lib/python2.7/dist-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.15rc1 (default, Apr 15 2018, 21:51:34) [GCC 7.3.0]

```

##### CONFIGURATION

./ansible.cfg
library        = /srv/ansible_modules
forks          = 50
log_path = ./ansible.log
gathering = explicit
host_key_checking = False
callback_whitelist = profile_tasks
timeout = 30
connect_timeout = 180
command_timeout = 60

##### OS / ENVIRONMENT

Ubuntu 18.04

##### STEPS TO REPRODUCE

Tested show clock command collection on 26 routers. Inventory and ansible.cfg files are the same for each playbook.
Run the following playbooks with callback_whitelist = profile_tasks enabled to check execution time.

playbook: test_network_cli-iosxr_command.yml
network_module: iosxr_command
connection_plugin: network_cli
```yaml
---

- name: collect show commands from devices
  hosts: test_group
  connection: network_cli

  tasks:

    - name: collect show commands
      no_log: True
      iosxr_command:
        commands: "show clock"
```

playbook: test_network_cli-ntc_show_command.yml
network_module: ntc_show_command
connection_plugin: network_cli
```yaml
---

- name: collect show commands from devices
  hosts: test_group
  connection: network_cli

  tasks:

    - name: collect show commands
      no_log: True
      ntc_show_command:
        platform: "cisco_xr"
        command: "show clock"
        host: "{{ inventory_hostname }}"
        username: "{{ ansible_user }}"
        password: "{{ ansible_ssh_pass }}"
        use_templates: False
```

playbook: test_netmiko_ssh-ntc_show_command.yml
network_module: ntc_show_command
connection_plugin: netmiko_ssh
```yaml
---

- name: collect show commands from devices
  hosts: test_group
  connection: local

  tasks:

    - name: collect show commands
      no_log: True
      ntc_show_command:
        platform: "cisco_xr"
        command: "show clock"
        host: "{{ inventory_hostname }}"
        username: "{{ ansible_user }}"
        password: "{{ ansible_ssh_pass }}"
        use_templates: False
```

##### EXPECTED RESULTS

Connections to routers should start simultaneously in parallel. Execution time should not increase linearly with number of hosts. 

##### ACTUAL RESULTS

**test1**
playbook: test_network_cli-iosxr_command.yml
-vvvv output link: https://gist.github.com/vaneuk/2efd3242fb75c4154025999e459ef922
execution time: **58.96s**

according to the outputs and 'show users' checks on routers ansible ssh sessions are created in sequential order, and not in parallel.

**test2**
playbook: test_network_cli-ntc_show_command.yml
-vvvv output link: https://gist.github.com/vaneuk/65dfcb40057842b2541a38bd0797cd64
execution time: **65.53s**

**test3**
playbook: test_netmiko_ssh-ntc_show_command.yml
-vvvv output link: https://gist.github.com/vaneuk/42ea715dcd9c72e3ffbdb4455878239c
execution time: **12.26s**

The only difference between playbooks in test2 and test3 are connection plugins. With plugin netmiko_ssh playbook runs 5x times faster.

```

```

**Comment by sivel** [comment]
I believe this issue has been resolved in https://github.com/ansible/ansible/pull/39223 which was subsequently included in the v2.5.3 release.

Can you upgrade to v2.5.3 or newer and validate whether this has been resolved?

needs_info

**Comment by ansibot** [comment]
cc @Qalthos @ganeshrn @gdpak @gundalow @privateip @rcarrillocruz @trishnaguha
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by vaneuk** [comment]
@sivel thanks for prompt response! Seems that you are right - I've tested on 2.5.5 and get 20 seconds on network_cli vs 12 seconds on netmiko_ssh.  It's much better :)

**Comment by sivel** [comment]
Thank you for testing.  If you have any further problems feel free to open a new issue.

If you have further questions please stop by IRC or the mailing list:

   * IRC: #ansible on irc.freenode.net
   * mailing list: https://groups.google.com/forum/#!forum/ansible-project
